### ISSUE TITLE
Feature Request: Prepend Query Metadata in SQL Comment

### ISSUE THREAD

**Issue by robinfrick** [metabase/metabase#2386]
Hey,

I'm back with a new feature request.
Metabase is a very powerful tool, but in the hands of the wrong people it can do more harm than good. We had the case this week that one of our colleagues put up a dashboard with fairly intensive queries in a loop where it would refresh every 30 seconds. That put a heavy load on the database as you can imagine.

It was kind of difficult for us to find the culprit and what was going on because we could see the queries and that they were originating from the user designated for metabase, but it was difficult to find out what questions it were and who was executing them.

Re:Dash has a nice solution to this they prepend a comment to every SQL Query stating among other things the question id and user like this:
`/* Username: Scheduled, Task ID: 66c28896-d5a0-4230-be02-ea0da7c8e92d, Query ID: 15, Queue: scheduled_queries, Query Hash: 59ad1137f86748f72eec0bd0a9acb8c1 */`
With this information I can quickly identify a query that is causing performance issues and also why/by whom it is being executed.
It would be really nice to have the same on Metabase as well.

Thanks for your awesome work :)

**Comment by salsakran** [comment]
This sounds like a pretty useful idea. Not sure there are analogs for druid, and mongo, but I think it'd be useful for DBs that support comments.

**Comment by salsakran** [comment]
So in terms of what we should pre-pend, I think something like the following would work for us
```/* Metabase:: user: <username> questionId: <questionId | "null" for ad hoc's> queryType: <native | gui>/```

Anything we'd want to add or remove from that?

**Comment by camsaul** [comment]
I think the most widely accepted SQL comment syntax is `--`

so probably more like

```
-- Metabase:: ...
```

**Comment by salsakran** [comment]
now that I think about it we should probably also add the query execution id

**Comment by camsaul** [comment]
query execution ID is different every time though, right? Would that make it harder to group together executions of the same query, since the comments would change every time?

**Comment by salsakran** [comment]
that would be in addition to the query id. 
that way if the person wants to debug a certain execution of a query , they have a tie between the query execution and the sql log. eg, hypothetically, if we were executing queries twice, it would show up as the same query execution id. Not likely to happen but good info to have around

**Comment by salsakran** [comment]
a +1 for this feature via the issue at  http://discourse.metabase.com/t/unknown-mysql-queries/370

If we had this, it would be trivial to trace whether those unknown queries came from Metabase.

**Comment by camsaul** [comment]
I've been playing around with this and it doesn't look like there's currently any way to do this with korma as the code base is right now

@agilliland's branch where we use korma to compile things to SQL but not for execution would make it easy to prepends to the generated string

**Comment by camsaul** [comment]
So I don't think we'll be able to include the ID of the card associated with a given query. We have an API endpoint `GET /api/dataset/card/:id` to fetch the data for a Card but as far as I can tell that endpoint is no longer used and the frontend just always hits `POST /api/dataset` directly. So the backend never has any information about what card might be associated with a query

Things we still can include:

*  "Metabase ::"
*  ID of user who executed the query
*  A hash of the query
*  UUID of associated `QueryExecution` 

Are those pieces of information enough?

**Comment by camsaul** [comment]
This is what I have it right now

```sql
-- Metabase:: userID: 2 executionID: 170cbab5-090b-4be5-b286-1168f9439b5a queryType: MBQL queryHash: 1005045833
SELECT "PUBLIC"."ORDERS"."USER_ID" AS "USER_ID", [...]
FROM "PUBLIC"."ORDERS" 
LIMIT 1
```

**Comment by camsaul** [comment]
Implemented by #2687

**Comment by salsakran** [comment]
re-opening due to bugs in the PR.

**Comment by CzechJiri** [comment]
This feature is typically **not** done using comments, but using connection metadata, which is a easy to query via separate column(s) and typically preserved in logs and performance reports


**ORACLE** uses concept of DBMS_APPLICATION_INFO with CLIENT_INFO / MODULE / ACTION which are for the most part preserved in AWR/ASM reports and take almost no time to setup/change per JDBC call
https://docs.oracle.com/cd/B28359_01/appdev.111/b28419/d_appinf.htm

**POSTGRES** has a concept of APPLICATION_NAME which can be set during or anytime after connection and is preserved in the logs 
https://www.postgresql.org/docs/9.5/static/runtime-config-logging.html#GUC-APPLICATION-NAME

**SQL SERVER** (it changed a lot since I was deep in it) has similar concept https://social.msdn.microsoft.com/Forums/sqlserver/en-US/6dacb63d-7eb4-449e-b747-a23ac67237ef/equallent-application-dbmsapplicationinfoin-sql-server?forum=sqlgetstarted

**Comment by salsakran** [comment]
@camsaul This needs to be fixed for 0.18 to be code complete.

**Comment by salsakran** [comment]
@CzechJiri That's super useful. Do you see any specific downside to comments? I.e. if we did both, do you foresee any problems

**Comment by CzechJiri** [comment]
the only potential drawback (this is really different for each database system) might be that some databases won't cache the execution plan or/and the result simply because the whole SQL is changing, but even if some DBs do this, I don't think this is an issue, because you are not running 3000 queries per second and really don't need to shave every millisecond 

it seems MS SQL might see potentially tiny impact https://www.brentozar.com/archive/2014/12/whitespace-comments-forced-parameterization-sql-server/

Oracle is a bit smarter does not care about comments https://docs.oracle.com/cd/B12037_01/server.101/b10759/sql_elements006.htm

**Comment by camsaul** [comment]
Fixed by #2743
